jwql/instrument_monitors/nircam_monitors/claw_monitor.py

```diff
@@ -0,0 +1,394 @@
+    for each filter/pupil combination. These stacks are then plotted
+    in on-sky orientation and the results are used to identify new
+    instances of claws - a scattered light effect seen in NIRCam
+    data. Background statistics are also stored for each individual
+    image, which are then plotted to track the NIRCam background
+    levels over time. Results are all saved to database tables.
+
+    Attributes
+    ----------
+    outfile : str
+        The name of the output plot for a given claw stack combination.
+
+    output_dir : str
+        Path into which claw stack plots will be placed.
+
+    output_dir_bkg : str
+        Path into which background trending plots will be placed.
+
+    query_start : float
+        MJD start date to use for querying MAST.
+
+    query_end : float
+        MJD end date to use for querying MAST.
+
+    wv : str
+        NIRCam channel for a given claw stack, either ``SW`` or ``LW``.
+
+    proposal : str
+        NIRCam proposal number for a given claw stack.
+
+    obs : str
+        NIRCam observation number for a given claw stack.
+
+    fltr : str
+        NIRCam filter used for a given claw stack.
+
+    pupil : str
+        NIRCam pupil used for a given claw stack.
+
+    detectors : str
+        The detectors used for a given claw stack combination.
+
+    files : numpy.ndarray
+        The names of the individual files belonging to a given claw stack combination.
+    """
+
+    def __init__(self):
+        """Initialize an instance of the ``ClawMonitor`` class.
+        """
+
+        # Define and setup the output directories for the claw and background plots.
+        self.output_dir = os.path.join(get_config()['outputs'], 'claw_monitor', 'claw_stacks')
+        ensure_dir_exists(self.output_dir)
+        self.output_dir_bkg = os.path.join(get_config()['outputs'], 'claw_monitor', 'backgrounds')
+        ensure_dir_exists(self.output_dir_bkg)
+
+        # Get the claw monitor database tables
+        self.query_table = eval('NIRCamClawQueryHistory')
+        self.stats_table = eval('NIRCamClawStats')
+
+    def make_background_plots(self):
+        """Makes plots of the background levels over time in NIRCam data.
+        """
+
+        # Get all of the background data.
+        query = session.query(NIRCamClawStats.filename, NIRCamClawStats.filter, NIRCamClawStats.pupil, NIRCamClawStats.detector,
+                              NIRCamClawStats.effexptm, NIRCamClawStats.expstart_mjd, NIRCamClawStats.entry_date, NIRCamClawStats.mean,
+                              NIRCamClawStats.median, NIRCamClawStats.frac_masked).all()
+        df_orig = pd.DataFrame(query, columns=['filename', 'filter', 'pupil', 'detector', 'effexptm', 'expstart_mjd',
+                                               'entry_date', 'mean', 'median', 'frac_masked'])
+        df_orig = df_orig.drop_duplicates(subset='filename', keep="last")  # remove any duplicate filename entries, keep the most recent
```

Comment:
This is a useful pandas method, I've never seen before. Going to make sure to keep this one on my mind for future projects!